tests/pubsub/testgossipsubheartbeat.nim

```diff
@@ -0,0 +1,305 @@
+{.used.}
+
+import std/[sequtils]
+import utils
+import ../../libp2p/protocols/pubsub/[gossipsub, mcache, peertable]
+import ../helpers
+
+suite "GossipSub Heartbeat":
+  teardown:
+    checkTrackers()
+
+  asyncTest "Mesh is rebalanced during heartbeat - pruning peers":
+    let
+      numberOfNodes = 10
+      topic = "foobar"
+      nodes = generateNodes(numberOfNodes, gossip = true).toGossipSub()
+      node0 = nodes[0]
+
+    startNodesAndDeferStop(nodes)
+
+    # Observe timestamps of received prune messages
+    var lastPrune = Moment.now()
+    for i in 0 ..< numberOfNodes:
+      let pubsubObserver = PubSubObserver(
+        onRecv: proc(peer: PubSubPeer, msgs: var RPCMsg) =
+          if msgs.control.isSome:
+            for msg in msgs.control.get.prune:
+              lastPrune = Moment.now()
+      )
+      nodes[i].addObserver(pubsubObserver)
+
+    # Nodes are connected to Node0
+    for i in 1 ..< numberOfNodes:
+      await connectNodes(node0, nodes[i])
+    subscribeAllNodes(nodes, topic, voidTopicHandler)
+    await waitForHeartbeat()
+    check:
+      node0.mesh[topic].len == 9
+
+    # When DValues of Node0 are updated to lower than defaults
+    let newDValues = some(
+      DValues(
+        dLow: some(2),
+        dHigh: some(4),
+        d: some(3),
+        dLazy: some(3),
+        dScore: some(2),
+        dOut: some(2),
+      )
+    )
+    node0.parameters.applyDValues(newDValues)
+
+    # Waiting 2 hearbeats to finish pruning (2 peers first and then 3)
+    # Comparing timestamps of last received prune to confirm heartbeat interval
+    await waitForHeartbeat()
+    let after1stHearbeat = lastPrune
+
+    await waitForHeartbeat()
+    let after2ndHearbeat = lastPrune
+
+    let measuredHeartbeat = after2ndHearbeat - after1stHearbeat
+    let heartbeatDiff = measuredHeartbeat - TEST_GOSSIPSUB_HEARTBEAT_INTERVAL
+
+    # Then mesh of Node0 is rebalanced and peers are pruned to adapt to new values
+    check:
+      node0.mesh[topic].len >= 2 and node0.mesh[topic].len <= 4
```

Comment:
`node0.mesh[topic].len >= 2 and node0.mesh[topic].len <= 4`

what are 2 and 4 here, `newDVales.dLow` and `newDVales.dHigh` ?